### PR TITLE
Http check revision

### PR DESCRIPTION
**Pull request by jdowni000**
Updated logic for http-check to iterate 10 times over to a URL and report accordingly to Kuberhealhty with a 80% passing margin

### CODE REVIEW

File: deploy/helm/kuberhealthy/templates/khcheck-pod-status.yaml [hunk 1]
```diff
@@ -27,6 +30,9 @@ spec:
           requests:
             cpu: 10m
             memory: 50Mi
+        securityContext:
```
Comment (integrii) [inline deploy/helm/kuberhealthy/templates/khcheck-pod-status.yaml:33]:
```suggestion
        {{- if .Values.securityContext.enabled }}
        securityContext:
```

File: deploy/helm/kuberhealthy/templates/khcheck-pod-status.yaml [hunk 2]
```diff
@@ -27,6 +30,9 @@ spec:
...
+          allowPrivilegeEscalation: {{ .Values.securityContext.allowPrivilegeEscalation }}
+          readOnlyRootFilesystem: {{ .Values.securityContext.readOnlyRootFilesystem }}
```
Comment (integrii) [inline deploy/helm/kuberhealthy/templates/khcheck-pod-status.yaml:35]:
```suggestion
          readOnlyRootFilesystem: {{ .Values.securityContext.readOnlyRootFilesystem }}
          {{- end}}
```

File: deploy/helm/kuberhealthy/templates/khcheck-pod-status.yaml [hunk 3]
```diff
@@ -8,6 +8,9 @@ spec:
   runInterval: 5m
   timeout: 15m
   podSpec:
```
Comment (integrii) [inline deploy/helm/kuberhealthy/templates/khcheck-pod-status.yaml:10]:
```suggestion
  podSpec:
      {{- if .Values.securityContext.enabled }}
```

File: deploy/helm/kuberhealthy/templates/khcheck-pod-status.yaml [hunk 4]
```diff
@@ -8,6 +8,9 @@ spec:
...
+    securityContext:
+      runAsUser: {{ .Values.securityContext.runAsUser }}
+      fsGroup: {{ .Values.securityContext.fsGroup }}
```
Comment (integrii) [inline deploy/helm/kuberhealthy/templates/khcheck-pod-status.yaml:13]:
```suggestion
      fsGroup: {{ .Values.securityContext.fsGroup }}
      {{- end}}
```

File: cmd/http-check/http-check.yaml [hunk 5]
```diff
@@ -10,8 +10,8 @@ spec:
   podSpec:
     containers:
     - name: http
-      image: kuberhealthy/http-check:v1.1.1
-      imagePullPolicy: IfNotPresent
+      image: kuberhealthy/http-check:v1.2.2
+      imagePullPolicy: Always
```
Comment (integrii) [inline cmd/http-check/http-check.yaml:14]:
```suggestion
      imagePullPolicy: IfNotPresent
```

File: cmd/http-check/main.go [hunk 6]
```diff
@@ -2,13 +2,14 @@ package main
 
 import (
 	"fmt"
+	"log"
```
Comment (jonnydawg) [inline cmd/http-check/main.go:5]: I think you should change this import back to logrus for consistency -- all of the other checks are using the package for logging: https://github.com/sirupsen/logrus

File: cmd/http-check/main.go [hunk 7]
```diff
@@ -29,30 +30,53 @@ func init() {
 }
 
 func main() {
-	log.Infoln("Beginning check.")
+	log.Println("Beginning check.")
+	checksRan := 0
+	checksPassed := 0
+	checksFailed := 0
+
+	ticker := time.NewTicker(time.Second)
```
Comment (jonnydawg) [inline cmd/http-check/main.go:38]: is it okay to spam 10 requests in 10 seconds to a URL? Would it get us blacklisted via robots.txt? https://support.google.com/webmasters/answer/6062608?hl=en

Comment (integrii) [reply cmd/http-check/main.go:38]: i doubt 10 requests in 10 seconds would get anyone banned from anything, but if we find that to be true we can adjust.

It is probably prudent to break the tries and seconds of delay into environment variables.  @jdowni000 can you implement please?  Be sure to include the new environment variables you create in the example specs.

File: cmd/http-check/main.go [hunk 8]
```diff
@@ -71,12 +71,12 @@ func main() {
...
 			log.Fatalln("error when reporting to kuberhealthy:", err.Error())
 		}
-		log.Println("Successfully reported to Kuberhealthy of failure")
+		log.Infoln("Successfully reported to Kuberhealthy of failure")
 	}
 
 	err := kh.ReportSuccess()
 	if err != nil {
-		log.Fatalln("error when reporting to kuberhealthy:", err.Error())
+		log.Infoln("error when reporting to kuberhealthy:", err.Error())
```
Comment (integrii) [inline cmd/http-check/main.go:79]: This used to Fatal, but now it only calls info, and then keeps executing.

Comment (jdowni000) [reply cmd/http-check/main.go]: I have corrected this back to log.Fatalln

File: cmd/http-check/main.go [hunk 9]
```diff
@@ -30,29 +31,52 @@ func init() {
...
-	log.Errorln(reportErr.Error())
-	kh.ReportFailure([]string{reportErr.Error()})
+	// Displays the results of 10 URL requests
+	log.Infoln(checksRan, "checks ran")
+	log.Infoln(checksPassed, "checks passed")
+	log.Infoln(checksFailed, "checks failed")
+
+	// Check to see if the 10 requests passed at 80% and reports to Kuberhealthy accordingly
+	if checksPassed < 8 {
+		reportErr := fmt.Errorf("unable to retrieve a http.StatusOK from " + checkURL + "check failed " + strconv.Itoa(checksFailed) + " out of 10 attempts")
```
Comment (jonnydawg) [inline cmd/http-check/main.go:69]: make sure to move the htt.StatusOK out of the first string -- it's a constant from the http package that should be put in the error

Comment (jdowni000) [reply cmd/http-check/main.go]: changed verbiage to "unable to retrieve a valid response from " as the checks failed is more than just http.StatusOk @jonnydawg

File: cmd/http-check/main.go (hunk 7 above)
Comment (jdowni000) [reply cmd/http-check/main.go]: I have added this logic with the latest commit @integrii

File: cmd/http-check/main.go [hunk 10]
```diff
@@ -23,23 +25,45 @@ func init() {
...
 
 func main() {
+
+	countInt, x := strconv.Atoi(count)
+	if x != nil {
+		log.Fatalln(x)
+	}
+
+	secondInt, y := strconv.Atoi(seconds)
+	if y != nil {
```
Comment (integrii) [inline cmd/http-check/main.go:52]: change `y` to `err` here

File: cmd/http-check/main.go [hunk 11]
```diff
@@ -23,23 +25,45 @@ func init() {
...
 	// If the URL does not begin with HTTP, exit.
 	if !strings.HasPrefix(checkURL, "http") {
 		log.Fatalln("Given URL does not declare a supported protocol. (http | https)")
 	}
 }
 
 func main() {
+
+	countInt, x := strconv.Atoi(count)
+	if x != nil {
```
Comment (integrii) [inline cmd/http-check/main.go:47]: change `x` to `err` here

File: cmd/http-check/main.go [hunk 12]
```diff
@@ -22,43 +26,94 @@ func init() {
...
+		checksRan++
+
+		if err != nil {
+			checksFailed++
+			log.Infoln("Failed to reach URL: ", checkURL)
+			continue
+		}
+
+		if r.StatusCode != http.StatusOK {
+			log.Infoln("Got a", r.StatusCode, "with a", http.MethodGet, "to", checkURL)
```
Comment (integrii) [inline cmd/http-check/main.go:90]: use `log.Errorln` here or even better yet, produce some summary of bad status codes and how many times they were seen, then report that to Kuberhealthy when failure occurs.

File: cmd/http-check/main.go [hunk 13]
```diff
@@ -22,43 +26,94 @@ func init() {
...
 	}
 
+	// Check that the COUNT environment variable is valid.
+	if len(count) == 0 {
+		log.Fatalln("Empty count value. Please update your COUNT environemnt variable.")
+	}
+
+	// Check that the SECONDS environment variable is valid.
+	if len(seconds) == 0 {
+		log.Fatalln("Empty seconds value. Please update your SECONDS environemnt variable.")
```
[the remark this answers is not in the document]
Comment (jdowni000) [reply cmd/http-check/main.go]: changed to  log.Errorln and an exit on the next line

File: cmd/http-check/main.go [hunk 14]
```diff
@@ -22,43 +26,94 @@ func init() {
...
-		}
-		os.Exit(0)
+		log.Fatalln(err)
 	}
 
-	// Check if the response status code is a 200.
-	if r.StatusCode == http.StatusOK {
+	secondInt, err := strconv.Atoi(seconds)
+	if err != nil {
+		log.Fatalln(err)
```
[the remark this answers is not in the document]
Comment (jdowni000) [reply cmd/http-check/main.go]: changed to  log.Errorln and an exit on the next line

File: cmd/http-check/main.go [hunk 15]
```diff
@@ -22,43 +26,94 @@ func init() {
...
+	}
+
+	passingFloat, err := strconv.ParseFloat(passing, 64)
+	if err != nil {
+		log.Fatalln(err)
```
[the remark this answers is not in the document]
Comment (jdowni000) [reply cmd/http-check/main.go]: changed to  log.Errorln and an exit on the next line

File: cmd/http-check/main.go [hunk 16]
```diff
@@ -22,43 +26,94 @@ func init() {
...
+	countInt, err := strconv.Atoi(count)
 	if err != nil {
-		reportErr := fmt.Errorf("error occurred performing a " + http.MethodGet + " to " + checkURL + ": " + err.Error())
-		log.Errorln(reportErr.Error())
-		reportFailureErr := kh.ReportFailure([]string{reportErr.Error()})
-		if reportFailureErr != nil {
-			log.Fatalln("error when reporting to kuberhealthy:", reportFailureErr.Error())
-		}
-		os.Exit(0)
+		log.Fatalln(err)
```
[the remark this answers is not in the document]
Comment (jdowni000) [reply cmd/http-check/main.go]: changed to  log.Errorln and an exit on the next line

File: cmd/http-check/main.go [hunk 17]
```diff
@@ -22,43 +26,94 @@ func init() {
...
+	log.Infoln(checksPassed, "checks passed")
+	log.Infoln(checksFailed, "checks failed")
+
+	// Check to see if the number of requests passed at 90% and reports to Kuberhealthy accordingly
+	if checksPassed < passInt {
+		reportErr := fmt.Errorf("unable to retrieve a valid response from " + checkURL + "check failed " + strconv.Itoa(checksFailed) + " out of " + strconv.Itoa(checksRan) + " attempts")
+		err := kh.ReportFailure([]string{reportErr.Error()})
 		if err != nil {
-			log.Fatalln("error when reporting to kuberhealthy:", reportSuccessErr.Error())
+			log.Fatalln("error when reporting to kuberhealthy:", err.Error())
```
[the remark this answers is not in the document]
Comment (jdowni000) [reply cmd/http-check/main.go]: I have left the fatal in

File: cmd/http-check/main.go [hunk 18]
```diff
@@ -22,43 +26,94 @@ func init() {
...
 	}
 
-	reportErr := fmt.Errorf("unable to retrieve a " + strconv.Itoa(http.StatusOK) + " from " + checkURL + " got a " + strconv.Itoa(r.StatusCode) + " instead")
-	log.Errorln(reportErr.Error())
-	reportFailureErr := kh.ReportFailure([]string{reportErr.Error()})
-	if reportFailureErr != nil {
-		log.Fatalln("error when reporting to kuberhealthy:", reportFailureErr.Error())
+	err = kh.ReportSuccess()
+	if err != nil {
+		log.Fatalln("error when reporting to kuberhealthy:", err.Error())
```
[the remark this answers is not in the document]
Comment (jdowni000) [reply cmd/http-check/main.go]: I have left the fatal in

File: cmd/http-check/main.go (hunk 12 above)
Comment (jdowni000) [reply cmd/http-check/main.go]: I changed this to an `log.Errorln`

File: cmd/http-check/main.go [hunk 19]
```diff
@@ -23,22 +23,26 @@ var (
 func init() {
 	// Check that the URL environment variable is valid.
 	if len(checkURL) == 0 {
-		log.Fatalln("Empty URL. Please update your CHECK_URL environment variable.")
+		log.Errorln("Empty URL. Please update your CHECK_URL environment variable.")
+		os.Exit(0)
```
Comment (integrii) [inline cmd/http-check/main.go:27]: If you `os.Exit(0)`, then kuberhealthy will report that the checker pod did not report in within its timeout.  Let's change these exits to report failure with the same message, then exit.

File: cmd/http-check/http-check.yaml [hunk 20]
```diff
@@ -9,17 +8,23 @@ spec:
...
+        image: kuberhealthy/http-check:v1.2.4
+        imagePullPolicy: IfNotPresent
+        env:
+          - name: CHECK_URL
+            value: "http://google.com"
+          - name: COUNT
+            value: "10"
+          - name: SECONDS
+            value: "1"
+          - name: PASSING
```
Comment (integrii) [inline cmd/http-check/http-check.yaml:21]: Can we switch this to `PASSING_PERCENT` and change the value to a whole number?

Comment (jdowni000) [reply cmd/http-check/http-check.yaml]: I have changed this to `PASSING_PERCENT`

File: cmd/http-check/main.go (hunk 19 above)
Comment (jdowni000) [reply cmd/http-check/main.go:28]: I have refactored these to report to kuberhealthy of the failure before exit

File: cmd/http-check/main.go [hunk 21]
```diff
@@ -66,7 +66,7 @@ func main() {
 
 	// if the passing count is empty, then default to 1
```
Comment (integrii) [inline cmd/http-check/main.go:67]:
```suggestion
	// if the passing count is empty, then default to 100%
```

Comment (jdowni000) [reply cmd/http-check/main.go:67]: yes it is @integrii everything checked out with this code for me locally

File: cmd/http-check/README.md [hunk 22]
```diff
@@ -15,16 +21,22 @@ metadata:
...
 spec:
-  runInterval: 2m
-  timeout: 6m
+  runInterval: 5m
+  timeout: 10m
   podSpec:
     containers:
     - name: http
-      image: kuberhealthy/http-check:v1.1.1
+      image: kuberhealthy/http-check:v1.2.3
```
Comment (integrii) [inline cmd/http-check/README.md:29]:
```suggestion
      image: kuberhealthy/http-check:v1.2.4
```

**Pull request closed without merging**